baus/subsidies.py

```diff
@@ -56,6 +56,80 @@ def coffer(policy, scenario):
+    res_units = residential_units.to_frame()
+    bldgs = buildings.to_frame()
+    parcels_geog = parcels_geography.to_frame()
+    taz_geog = taz_geography.to_frame()
+
+    res_units = res_units.merge(bldgs[['parcel_id']], left_on='building_id', 
+                                right_index=True, how='left').\
+        merge(parcels_geog[['gg_id', 'sesit_id', 'tra_id',
+              'juris']], left_on='parcel_id', right_index=True, how='left').\
+        merge(taz_geog, left_on='zone_id', right_index=True, how='left')
+
+    s = policy["unit_preservation"]["settings"]
+
+    # only preserve units that are not already deed-restricted
+    res_units = res_units.loc[res_units.deed_restricted != 1]
+
+    # initialize list of units to mark deed restricted
+    dr_units = []
+    
+    # apply deed-restriced units by geography (county here)
+    for geog, value in s.items(): 
+
+        # apply deed-restriced units by filters within each geography 
+        l = ['first', 'second', 'third', 'fourth']
+        for item in l:
+
+            if value[item+"_unit_filter"] is None or \
+                value[item+"_unit_target"] is None:
+                    continue
+            
+            filter_nm = value[item+"_unit_filter"]
+            unit_target = value[item+"_unit_target"]
+
+            # exclude units that have been preserved through this loop
+            res_units = res_units[~res_units.index.isin(dr_units)]
+
+            # subset units to the geography
+            geography = policy["unit_preservation"]["geography"]
+            geog_units = res_units.loc[res_units[geography] == geog]
+            # subset units to the filters within the geography
+            filter_units = geog_units.query(filter_nm)
+
+            # pull a random set of units based on the target except in cases
+            # where there aren't enough units in the filtered geography or
+            # they're already marked as deed restricted
+            if len(filter_units) == 0:
+                dr_units_set = []
+                print("%s %s: target is %d but no units are available" % 
+                      (geog, filter_nm, unit_target))
+            elif unit_target > len(filter_units):
+                 dr_units_set = filter_units.index
+                 print("%s %s: target is %d but only %d units are available" % 
+                      (geog, filter_nm, unit_target, len(filter_units)))
+            else:
+                dr_units_set = np.random.choice(filter_units.index, 
+                                                unit_target, replace=False)
+
+            dr_units.extend(dr_units_set)
+
+    # mark units as deed restriced in residential units table
+    residential_units = residential_units.to_frame()
+    residential_units.loc[residential_units.index.isin(dr_units), 
+                          'deed_restricted'] = 1
+    orca.add_table("residential_units", residential_units)
```

Comment:
In addition to updating the "**residential_units**" table, I think the function should also update the `deed_restricted_units` 
 column of the "**buildings**" table. Otherwise, there will be fewer deed_restricted_units in the buildings table than in the "**residential_units**" table, right?

Moreover, I'm not sure if it is necessary to add a new column to the "**buildings**" table called `preserved_units`. Without this column, are we still able to separate preserved_units and newly created subsidized_units? Right now `deed_restricted_units - inclusionary_units = subsidized_units`, but when we add the preservation strategy, `deed_restricted_units - inclusionary_units = created_subsidized_units + preserved_units`, right? We can say that "preserved" units are also "subsidized", so we'll report the results of the two strategies together.

Reply:
@yuqiww I replied to this last week but I guess it didn't save. Yes, I was in an hlcm-level mindset- I'll get these into the buildings table so they show up in the summaries and other deed-restricted accounting. Thanks!

Reply:
I'll also add them in a way that allows us to parse them! If they stay "deed-restricted", then `deed_restricted_units - inclusionary_units - subsidized_units = preserved, existing, or pipeline`. So that's one incremental way to do it, if you consider existing and preserved similar. Otherwise "subsidized" would also be partially useful. And a new "preserved' column would be ideal.

Reply:
I agree that a new "preserved_units" column is the best, after all, this is a stand-alone strategy.